examples/keras_recipes/approximating_non_function_mappings.py

```diff
@@ -0,0 +1,424 @@
+import matplotlib.pyplot as plt
+from collections import OrderedDict
+import math
+import random
+from keras_core import callbacks
+import keras_core
+import tensorflow as tf
+from keras_core import layers
+from keras_core import optimizers
+from tensorflow_probability import distributions as tfd
+
+"""
+Next, lets generate a noisy spiral that we're going to attempt to approximate.
+I've defined a few functions below to do this:
+"""
+
+
+def normalize(x):
+    return (x - np.min(x)) / (np.max(x) - np.min(x))
+
+
+def create_noisy_spiral(n, max_jitter=0.2, size_range=[5, 10]):
+    R_X = random.choice(range(size_range[0], size_range[1], 10))
+    R_Y = random.choice(range(size_range[0], size_range[1], 10))
+    X_0 = random.random() * 100 - 50
+    Y_0 = random.random() * 100 - 50
+
+    randomized_jitter = lambda: random.uniform(-max_jitter / 2, max_jitter / 2)
+    x = lambda _t: _t / 360.0 * math.cos(_t / 90.0 * math.pi) * R_X + X_0
+    y = lambda _t: _t / 360.0 * math.sin(_t / 90.0 * math.pi) * R_Y + Y_0
+
+    out = np.zeros([n, 2])
+
+    for i in range(n):
+        t = 360.0 / n * i
+        out[i, 0] = x(t) + randomized_jitter()
+        out[i, 1] = y(t) + randomized_jitter()
+    out = out.astype("float32")
+    return (normalize(out[:, 0]) * 10) - 5, (normalize(out[:, 1]) * 10) - 5
+
+
+"""
+Next, lets invoke this function many times to construct a sample dataset:
+"""
+
+xs = []
+ys = []
+
+for _ in range(10):
+    x, y = create_noisy_spiral(1000)
+    xs.append(x)
+    ys.append(y)
+
+x = np.concatenate(xs, axis=0)
+y = np.concatenate(ys, axis=0)
+x = np.expand_dims(x, axis=1)
+
+plt.scatter(x, y)
+plt.show()
+
+"""
+As you can see, there's multiple possible values for Y with respect to a given
+X.  Normal neural networks will simply learn the mean of these points with
+respect to geometric space.
+
+We can quickly show this with a simple linear model:
+"""
+
+N_HIDDEN = 128
+
+model = keras_core.Sequential(
+    [
+        layers.Dense(N_HIDDEN, activation="relu"),
+        layers.Dense(N_HIDDEN, activation="relu"),
+        layers.Dense(1, activation="relu"),
+    ]
+)
+
+"""
+Let's use mean squared error as well as the adam optimizer.
+These tend to be reasonable prototyping choices:
+"""
+
+model.compile(optimizer="adam", loss="mse")
+
+"""We can fit this model quite easy"""
+
+model.fit(
+    x,
+    y,
+    epochs=300,
+    batch_size=128,
+    validation_split=0.15,
+    callbacks=[callbacks.EarlyStopping(monitor="val_loss", patience=5)],
+)
+
+"""
+And let's check out the result:
+"""
+
+y_pred = model.predict(x)
+
+"""
+As expected, the model learns the geometric mean of all points in `y` for a
+given `x`.
+"""
+
+plt.scatter(x, y)
+plt.scatter(x, y_pred)
+plt.show()
+
+"""
+
+## Mixture Density Networks
+
+Mixture Density networks can alleviate this problem.
+A Mixture density is a class of complicated densities expressible in terms of simpler densities.
+They are effectively the sum of a ton of probability distributions.
```

Comment:
effectively, a mixture density is the sum of various probability distributions.  By summing various distributions, mixture densitry distributions can effectively model arbitrarily complex distributions.